merkle/compact/tree.go

```diff
@@ -210,3 +186,22 @@ func (t *Tree) getNodes() [][]byte {
 	}
 	return n
 }
+
+// TreeNodes returns the list of node IDs that comprise a compact tree, in the
+// same order they are used in compact.Tree and compact.Range, i.e. ordered
+// from upper to lower levels.
+func TreeNodes(size uint64) []NodeID {
+	ids := make([]NodeID, 0, bits.OnesCount64(size))
+	// Iterate over perfect subtrees along the right border of the tree. Those
+	// correspond to the bits of the tree size that are set to one.
+	for sz := size; sz != 0; sz &= sz - 1 {
+		level := uint(bits.TrailingZeros64(sz))
+		index := (sz - 1) >> level
+		ids = append(ids, NewNodeID(level, index))
+	}
+	// Note: Right border nodes of compact.Range are ordered from root to leaves.
+	for i, j := 0, len(ids)-1; i < j; i, j = i+1, j-1 {
+		ids[i], ids[j] = ids[j], ids[i]
```

Comment:
Just a thought but we could add the other ordering to Range if reversal is going to crop up a lot.

Reply:
Good thought, we'll see if/when another use-case appears. That wouldn't be a trivial change though. `Range` implementation has a very specific ordering: `hashes` slice contains 2 **stacks** of hashes - the front one is ordered bottom-to-top, the back one is ordered top-to-bottom. Stack ordering comes handy when 2 ranges are merged, so that the slice is mutated with minimal copies/reallocations.